DIPs/DIP1xxx-rn.md

```diff
@@ -0,0 +1,801 @@
+to fix all the postblit issues.
+
+### Overview of this(this)
+
+The postblit function is a non-overloadable, non-qualifiable
+function. However, the compiler does not reject the
+following code:
+
+```D
+struct A { this(this) const {} }
+struct B { this(this) immutable {} }
+struct C { this(this) shared {} }
+```
+
+Since the semantics of the postblit in the presence of qualifiers was
+not defined and most likely not intended, this led to a series of problems:
+
+* `const` postblits are not able to modify any fields in the destination
+* `immutable` postblits never get called (resulting in compilation errors)
+* `shared` postblits cannot guarantee atomicity while blitting the fields
+
+#### `const`/`immutable` postblits
+
+The solution for `const` and `immutable` postblits is to type check them as normal
+constructors where the first assignment of a member is treated as an initialization
+and subsequent assignments are treated as modifications. This is problematic because
+after the blitting phase, the destination object is no longer in its initial state
+and subsequent assignments to its fields will be regarded as modifications making
+it impossible to construct `immutable`/`const` objects in the postblit. In addition,
+it is possible for multiple postblits to modify the same field. Consider:
+
+```D
+struct A
+{
+    immutable int a;
+    this(this)
+    {
+        this.a += 2;  // modifying immutable or ok?
+    }
+}
+
+struct B
+{
+    A a;
+    this(this)
+    {
+        this.a.a += 2;  // modifying immutable error or ok?
+    }
+}
+
+void main()
+{
+    B b = B(A(7));
+    B c = b;
+}
+```
+
+When `B c = b;` is encountered, the following actions are taken:
+
+1. `b`'s fields are blitted to `c`
+2. A's postblit is called
+3. B's postblit is called
+
+After `step 1`, the object `c` has the exact contents as `b` but it is not
+initialized (the postblits still need to fix it) nor uninitialized (the field
+`B.a` does not have its initial value). From a type checking perspective
+this is a problem because the assignment inside A's postblit is breaking immutability.
+This makes it impossible to postblit objects that have `immutable`/`const` fields.
+To alleviate this problem we can consider that after the blitting phase the object is
+in a raw state, therefore uninitialized; this way, the first assignment of `B.a.a`
+is considered an initialization. However, after this step, the field
+`B.a.a` is considered initialized, therefore how is the assignment inside B's postblit
+supposed to be type checked ? Is it breaking immutability or should
+it be legal? Indeed it is breaking immutability because it is changing an immutable
+value, however being part of initialization (remember that `c` is initialized only
+after all the postblits are ran) it should be legal, thus weakening the immutability
+concept and creating a different strategy from the one that normal constructors
+implement.
+
+#### Shared postblits
+
```

Comment:
The whole rationale behind `shared postblits` and/or copy constructors is based on assumptions that are, at least in my opinion, misguided. What, exactly, would be a purpose of a `shared` copy of a `shared` object? `shared` types, by definition, cannot, and shall not, be copyable. They're *shared* unique instances. A `shared` struct is either a global, or an allocated instance passed by reference. If you can freely make a copy, then sharing makes absolutely no sense.
The only contexts where it would make sense is acquiring an un-`shared` copy from a `shared` instance, or instantiating a `shared` instance from an un-`shared` one, both are beyond simple copying, they're type conversions, procedures which in the general case can only be reasoned about by user code.
Therefore, copy constructors/postblits for `shared` types are best made not allowed by the language wholesale, leaving any special treatment (i.e. conversion to/from `shared`) up to user code. This particular part of the DIP is trying to reason about a problem that is ill-treated (poorly specified) by the language, and ultimately would become a wasted effort.